language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+
+    /// Runs the prologue for the given transaction.
+    fn run_prologue<S: MoveStorage>(
+        &self,
+        session: &mut Session<S>,
+        transaction: &SignatureCheckedTransaction,
+        currency_code: &Identifier,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(), VMStatus>;
+}
+
+/// Validate a signed transaction by performing the following:
+/// 1. Check the signature(s) included in the signed transaction
+/// 2. Check that the transaction is allowed in the context provided by the `adapter`
+/// 3. Run the prologue to perform additional on-chain checks
+/// The returned `VMValidatorResult` will have status `None` and if all checks succeeded
+/// and `Some(DiscardedVMStatus)` otherwise.
+pub fn validate_signed_transaction<A: VMAdapter>(
+    adapter: &A,
+    transaction: SignedTransaction,
+    state_view: &dyn StateView,
+) -> VMValidatorResult {
+    let _timer = TXN_VALIDATION_SECONDS.start_timer();
+    let txn_sender = transaction.sender();
+    let log_context = AdapterLogSchema::new(state_view.id(), 0);
+
+    let txn = if let Ok(t) = A::check_signature(transaction) {
+        t
+    } else {
+        return VMValidatorResult::error(StatusCode::INVALID_SIGNATURE);
+    };
+
+    let remote_cache = StateViewCache::new(state_view);
+    let account_role = get_account_role(txn_sender, &remote_cache);
+    let mut session = adapter.new_session(&remote_cache);
+
+    let (status, normalized_gas_price) = match validate_signature_checked_transaction(
+        adapter,
+        &mut session,
+        &txn,
+        &remote_cache,
+        true,
+        &log_context,
+    ) {
+        Ok((price, _)) => (None, price),
+        Err(err) => (Some(err.status_code()), 0),
+    };
+
+    // Increment the counter for transactions verified.
+    let counter_label = match status {
+        None => "success",
+        Some(_) => "failure",
+    };
+    TRANSACTIONS_VALIDATED
+        .with_label_values(&[counter_label])
+        .inc();
+
+    VMValidatorResult::new(status, normalized_gas_price, account_role)
+}
+
+fn get_account_role(sender: AccountAddress, remote_cache: &StateViewCache) -> GovernanceRole {
+    let role_access_path = create_access_path(sender, RoleId::struct_tag());
+    if let Ok(Some(blob)) = remote_cache.get(&role_access_path) {
+        return bcs::from_bytes::<account_config::RoleId>(&blob)
+            .map(|role_id| GovernanceRole::from_role_id(role_id.role_id()))
+            .unwrap_or(GovernanceRole::NonGovernanceRole);
+    }
+    GovernanceRole::NonGovernanceRole
+}
+
+pub(crate) fn validate_signature_checked_transaction<S: MoveStorage, A: VMAdapter>(
+    adapter: &A,
+    mut session: &mut Session<S>,
+    transaction: &SignatureCheckedTransaction,
+    remote_cache: &S,
+    allow_too_new: bool,
+    log_context: &AdapterLogSchema,
+) -> Result<(u64, Identifier), VMStatus> {
+    if transaction.is_multi_agent() && !adapter.allows_multi_agent()? {
```

Comment:
It would be great if we could get `is_multi_agent` and `allows_multi_agent` out of the generic API and do a check inside `DiemVM` instead if that's feasible... would make the trait that much smaller. 

Maybe a more generic API like `is_transaction_supported(t: SignatureCheckerTransaction)` or `is_transaction_supported(t: RawTransaction)` would be useful for this sort of tx feature gating.

Reply:
Yup! Will change this and make the trait smaller.

Reply:
Is this a config/setting that you don't anticipate being needed in other scenarios?

Reply:
I think something like `is_transaction_supported` will be required by all adapters because of the deployment model for validator software/the framework. But I'm not sure whether it belongs in the `VMAdapter` trait, adapter_common, or individual customizations on top of adapter_common (I think probably the latter)?

Reply:
I have changed `adapter.allows_multi_agent()` to a more general `adapter.check_transaction_format(txn)`. And DiemVM's impl checks multi agent and duplicate signers.